### PR TITLE
Break out GrpcSender, GrpcSenderProvider

### LINKED ISSUE #5572: Should we add GrpcSender (and SPI)?
In #5505 / #5533 we added a HttpSender abstraction and SPI interface for providing different implementations. A big motivator for this was to allow `opentelemetry-exporter-otlp` consumers to avoid the okhttp dependency, which can be [problematic](#4777) in certain applications. 

#5557 provides an alternative implementation based on the jdk http client, which is great, but may not go far enough to solving the problem:`opentelemetry-exporter-otlp` bundles the grpc and http/protobuf variants of the OTLP exporters. The HttpSender / SPI solution remove the okhttp dependency for http/protobuf, but [grpc still depends on okhttp](https://github.com/open-telemetry/opentelemetry-java/blob/9fb64037175b5dd9ecdc640faf3f0328447d0733/exporters/otlp/all/build.gradle.kts#L23). 

Consumers can work around this by excluding `"com.squareup.okhttp3:okhttp` in maven / gradle, but then they must be careful to not use the grpc exporters. 

A more elegant solution would be to:
- Provide a corresponding GrpcSender abstraction and SPI
- Rescope the new artifact `opentelemetry-exporters-http-sender-okhttp` to include both HttpSender and GrpcSender SPI implementations, and rename it to something generic like `opentelemetry-exporter-sender-okhttp`. 
- Add a new artifact `opentelemetry-exporters-grpc-sender-upstream` containing the GrpcSender abstraction driven by the upstream grpc libraries. 

We would then have to document that consumers of `opentelemetry-exporter-otlp` need to be aware: 
- that the artifact contains both grpc and http/protobuf implementations
- by default, `opentelemetry-exporters-sender-okhttp` is included as a transitive dependency, and contains implementations of both
- if you want to opt out of okhttp, you need to add dependencies on at least one of `opentelemetry-exporters-http-sender-jdk` /  `opentelemetry-exporters-grpc-sender-upstream` and be aware of the requirements of the exporter protocol you intend on using (grpc or http/protobuf)

### PR DESCRIPTION
**Pull request by jack-berg**
Resolves #5572.

This PR touches a lot of files, and I'm not sure it can be reasonably split up. The HttpSender version was able to be split up because there was initially only an okhttp implementation - the jdk implementation was green field. In the GrpcSender case, there are two implementations (okhttp and grpc-upstream) which need to be supported out of the box. On the bright side, the patterns here mirror the HttpSender implementation so there's very little new material.

Some details about the implementation:
- The okhttp implementation is rolled into the existing `opentelemetry-exporter-sender-okhttp` artifact. The GrpcSender implementation doesn't require any new artifacts so better to package it with HttpSender to reduce artifact sprall.
- The grpc upstream implementation is publishe din a new artifact named `opentelemetry-exporter-sender-grpc-upstream`.
- To use the grpc upstream, you need to:
  - Add a dependency on `opentelemetry-exporter-sender-grpc-upstream`
  - Exclude the default dependency on `opentelemetry-exporter-sender-okhttp`
  - Set the channel manually using `OtlpGrpc{Signal}ExporterBuilder#setChannel(ManagedChannel)`

On that last note, its technically redundant to require users to include the grpc upstream artifact AND explicitly set the managed channel. We might consider adjusting the behavior to use grpc upstream if its the only GrpcSenderProvider implementatino present. Doing so would get us back into the business of configuring `ManagedChannel` instances from configuration parameters, which we stopped doing when we deprecated `#setChannel(ManagedChannel)`.

### CODE REVIEW

**Review by mateuszrzeszutek — APPROVED** [review]

File: exporters/otlp/all/build.gradle.kts [hunk 1]
```diff
@@ -45,27 +43,60 @@ testing {
...
+      targets {
+        all {
+          testTask {
+            systemProperty(
+              "io.opentelemetry.exporter.internal.grpc.GrpcSenderProvider",
+              "io.opentelemetry.exporter.sender.grpc.upstream.internal.UpstreamGrpcSenderProvider"
+            )
+          }
+        }
+      }
```
[the remark this answers is not in the document]
Comment (jack-berg) [reply exporters/otlp/all/build.gradle.kts:99]: This test uses the okhttp version of the upstream grpc library (i.e. io.grpc:grpc-okhttp). All the upstream grpc library implementations use the `GrpcSender` / `GrpcSenderProvider`.

**Comment by trask** [comment]
I don't love the name `opentelemetry-exporter-sender-grpc-upstream`, but I'm not sure I have any better ideas:

* `opentelemetry-exporter-sender-grpc-lib`
* `opentelemetry-exporter-sender-grpc-api`
* `opentelemetry-exporter-sender-grpc-java`
* `opentelemetry-exporter-sender-grpc-dep`

**Comment by jack-berg** [comment]
I don't mind `opentelemetry-exporter-sender-grpc-lib`. 

What about `opentelemetry-exporter-sender-grpc-transport`? "transport" is the terminology [used by the grpc library](https://github.com/grpc/grpc-java#transport).  

Or `opentelemetry-exporter-sender-grpc-channel`? "channel" is the abstraction over "transport", and may be a better description of what this new sender artifact contains, since its agnostic about _which_ underlying grpc transport is used.

**Comment by trask** [comment]
> What about `opentelemetry-exporter-sender-grpc-transport`? "transport" is the terminology [used by the grpc library](https://github.com/grpc/grpc-java#transport).

"grpc-transport" sounds like it could also describe our hand-rolled grpc transport layer

> Or `opentelemetry-exporter-sender-grpc-channel`? "channel" is the abstraction over "transport", and may be a better description of what this new sender artifact contains, since its agnostic about _which_ underlying grpc transport is used.

it may be a little subtle for some people, but the whole thing is kinda subtle

(similar, not necessarily better) `opentelemetry-exporter-sender-grpc-managed-channel` meaning, the only reason to use this is if you need access to the ManagedChannel(?)

**Comment by jack-berg** [comment]
`opentelemetry-exporter-sender-grpc-managed-channel` is verbose but clear that its an implementation that utilizes ManagedChannel to execute the export requests.

If others agree I'll update the artifact name to this^^.

**Comment by jkwatson** [comment]
> @jkwatson any thoughts on this?

I haven't spent a lot of time thinking about this...but my only thought currently is that this adds 3 new published modules, none of which have any public-facing APIs. That feels ... odd to me ... and I haven't really thought through the pros and cons of that.

**Comment by jack-berg** [comment]
It is a bit odd, but the alternative is using gradle exclude syntax to manage transitive dependencies not part of `io.opentelemetry.*` to achieve the same effect. I.e. if you want to use the java 11+ http client version, then exclude _this_ set of okhttp dependencies. 

I think this is the cleanest way to handle the sprawl of use cases we want to support as dictated by the java ecosystem. And with the `HttpSender` / `GrpcSender` SPI setup, we've added some future proofing against having to make significant refactors to accommodate whatever comes next, since users who are unsatisfied can provide their own implementation in a pinch. 

I do think that this is complex enough where we need some documentation. I'm imagining a list of use cases like and corresponding dependency management instruments, like: "If using java 11+ and you prefer to avoid extra dependencies, use the following configuration: ...". I can take an AI to add this kind of documentation to opentelemetry.io to help alleviate confusion.

**Comment by jkwatson** [comment]
Do we want to make these new modules alpha until we've settled on naming? Or are we confident enough that the names are "good enough" to go with?

**Comment by jack-berg** [comment]
I think the names are good enough. Marking the artifacts as alpha indicates that the've taken a step back in stability / reliability, which isn't the case. I will note that the `opentelemetry-exporter-sender-jdk` artifact _is_ alpha because it still needs some time in the wild before it can earn the stable designation. So the stable artifacts are `opentelemetry-exporter-sender-okhttp` and `opentelemetry-exporter-sender-grpc-managed-channel`.

**Pull request merged**